### PR TITLE
upgrade a few dependencies

### PR DESCRIPTION
**Pull request by ordian**
polkadot companion: https://github.com/paritytech/polkadot/pull/2205

### CODE REVIEW

**Review by andresilva — APPROVED** [review]
I haven't followed up on the latest `kvdb-*` changes, looking at the changelog there doesn't seem to be anything critical there so this is probably safe but maybe it still makes sense to do a burn-in for this PR.

Also could you check if polkadot has any direct depencies on the crates you updated (I'm almost certain it does) and update them there as well?

**Comment by ordian** [comment]
@bkchr did impl-trait-for-tuples 0.2 broke `const`?

**Review by bkchr — APPROVED** [review]

**Comment by ordian** [comment]
I'm not sure what would be the best way to fix web-wasm check.

The problem is that `primitive-types` indirectly pulls `getrandom` 0.2, which doesn't compile to `wasm32-unknown-unknown` unless `js` feature [is enabled](https://docs.rs/getrandom/0.2.1/getrandom/#webassembly-support).

I could add `getrandom` as an optional dependency to all crates that we check here https://github.com/paritytech/substrate/blob/71e47d792ff9997e742b8f90e4fb3a8397f502c3/.gitlab-ci.yml#L304-L317
but maybe there is a cleaner way?

**Comment by bkchr** [comment]
Did we used `getrandom` before?

**Comment by ordian** [comment]
> Did we used `getrandom` before?

yes, but version 0.1, this is a fallout from rand 0.8 upgrade in `fixed-hash`

**Comment by bkchr** [comment]
We could create a meta crate. The "test-wasm-browser-compilation" and all crates we want to test for wasm, we make a dependency of this crate. This way we can also add the getrandom and activate the feature.

**Comment by ordian** [comment]
> We could create a meta crate. The "test-wasm-browser-compilation" and all crates we want to test for wasm, we make a dependency of this crate. This way we can also add the getrandom and activate the feature.

We already test `bin/node/cli` this way, it includes most of these crates. I think we wanted to test they compile in separation?

File: client/consensus/aura/Cargo.toml [hunk 1]
```diff
@@ -37,6 +37,7 @@ sp-timestamp = { version = "2.0.0", path = "../../../primitives/timestamp" }
 sp-keystore = { version = "0.8.0", path = "../../../primitives/keystore" }
 sc-telemetry = { version = "2.0.0", path = "../../telemetry" }
 prometheus-endpoint = { package = "substrate-prometheus-endpoint", path = "../../../utils/prometheus", version = "0.8.0"}
+activate-wasm-bindgen-features = { version = "0.1.0", path = "../../../utils/activate-wasm-bindgen-features", optional = true }
```
Comment (bkchr) [inline client/consensus/aura/Cargo.toml:40]: Please no. You add a dependency to a crate that is set to `publish = no`.

Than just remove aura completely from the wasm test compilation.

Or just build the node-template.

File: utils/browser/Cargo.toml [hunk 2]
```diff
@@ -26,11 +26,6 @@ kvdb-web = "0.8.0"
...
-sc-chain-spec = { path = "../../client/chain-spec", version = "2.0.0"}
-
-# Imported just for the `wasm-bindgen` feature
-getrandom = { version = "0.2", features = ["js"] }
-rand = { version = "0.7", features = ["wasm-bindgen"] }
-futures-timer = { version = "3.0.1", features = ["wasm-bindgen"]}
-chrono = { version = "0.4", features = ["wasmbind"] }
+sc-network = { path = "../../client/network", version = "0.8.0" }
+sc-chain-spec = { path = "../../client/chain-spec", version = "2.0.0" }
+activate-wasm-bindgen-features = { version = "0.1.0", path = "../activate-wasm-bindgen-features" }
```
Comment (bkchr) [inline utils/browser/Cargo.toml:31]: Same here.

**Review by bkchr — CHANGES_REQUESTED** [review]

File: client/consensus/aura/Cargo.toml (hunk 1 above)
Comment (ordian) [reply client/consensus/aura/Cargo.toml]: > Or just build the node-template.

this requires some shenanigans to make it work, I'd like to leave it out of scope of this PR

went with a simple approach of adding one optional dependency to aura

**Review by bkchr — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Comment by vgeddes** [comment]
@ordian Just an FYI, the problem with `getrandom` is seemingly causing the build for our WASM runtime to break when we try to upgrade to the latest Substrate.
```
vincent@oryx:~/code/polkadot-ethereum/parachain (vincent/update-dependencies)$ cargo build
...
  error: target is not supported, for more information see: https://docs.rs/getrandom/#unsupported-targets
     --> /home/vincent/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.1/src/lib.rs:214:9
      |
  214 | /         compile_error!("target is not supported, for more information see: \
  215 | |                         https://docs.rs/getrandom/#unsupported-targets");
      | |_________________________________________________________________________^

  error[E0433]: failed to resolve: use of undeclared crate or module `imp`
     --> /home/vincent/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.1/src/lib.rs:236:5
      |
  236 |     imp::getrandom_inner(dest)
      |     ^^^ use of undeclared crate or module `imp`

  error: aborting due to 2 previous errors
```

Not sure how to go about fixing it. I tried to add the `js` feature with the following but it was ignored it seems:
```
[patch.crates-io]
getrandom = { git = "https://github.com/rust-random/getrandom.git", tag = "v0.2.1", features = ["js"] }
```

We also pull in `getrandom v0.2.1` via `ethereum-types` which is a dependency of several of our crates.

**Comment by ordian** [comment]
I don't think this patch will work, you'd have to patch all crates that pull `getrandom` indirectly if you go this way.
Have you tried adding this to the `Cargo.toml` crate that is pulling it?
```toml
getrandom = { version = "0.2", features = ["js"], optional = true }
```
See https://github.com/paritytech/substrate/pull/7831/commits/d245f58488125ab5a7a63be537b04a6eb0c9f3f7.

**Comment by ordian** [comment]
Actually, it seems that a `rand` feature of `fixed-hash` is leaking where it shouldn't. Let me try to find a source.

**Comment by vgeddes** [comment]
> I don't think this patch will work, you'd have to patch all crates that pull `getrandom` indirectly if you go this way.
> Have you tried adding this to the `Cargo.toml` crate that is failing to build?
> 

Thanks! that worked for me.

**Comment by ordian** [comment]
> > I don't think this patch will work, you'd have to patch all crates that pull `getrandom` indirectly if you go this way.
> > Have you tried adding this to the `Cargo.toml` crate that is failing to build?
> 
> Thanks! that worked for me.

Note that this is more of a temporary hack, rather than a fix. The actual fix would be to identify why `getrandom` is pulled in the first place (I don't think it should, looks like `std`/feature leakage of some sort).

**Comment by ordian** [comment]
@vgeddes FYI 
* https://github.com/paritytech/parity-common/pull/503
* https://github.com/rust-random/rand/pull/1086

**Comment by vgeddes** [comment]
Awesome! yeah I think the `serialize` feature for `ethereum-types` was also causing my WASM build to break yesterday (after fixing the getrandom issue).  Glad the issue has been confirmed. Luckily I didnt need the `serialize` feature so I removed it.

Upgrading our cargo resolver to v2 may have helped things too, though not sure.

**Pull request merged**